File: Java/src/main/java/com/gildedrose/ItemFactory.java
package com.gildedrose;

import static com.gildedrose.ItemType.fromName;

public class ItemFactory {
    public static Item createItem(String name, int sellIn, int quality) {
        var itemType = fromName(name);

        return switch (itemType) {
            case AgedBrie -> new AgedBrieItem(name, sellIn, quality);
            case BackstagePass -> new BackstagePassItem(name, sellIn, quality);
            case Sulfuras -> new SulfurasItem(name, sellIn, quality);
            case Normal -> new NormalItem(name, sellIn, quality);
        };
    }
}

File: src/Items/Factory/ItemFactory.php
<?php

declare(strict_types=1);

namespace GildedRose\Items\Factory;

use GildedRose\Item;
use GildedRose\Items\AgedBrieItem;
use GildedRose\Items\BackstagePassItem;
use GildedRose\Items\ConjuredItem;
use GildedRose\Items\Interface\ItemInterface;
use GildedRose\Items\NormalItem;
use GildedRose\Items\SulfurasItem;

class ItemFactory
{
    public const AGED_BRIE_ITEM = 'Aged Brie';

    public const BACKSTAGE_ITEM = 'Backstage';

    public const SULFURAS_ITEM = 'Sulfuras';

    public const CONJURED_ITEM = 'Conjured';

    public static function createItem(Item $item): ItemInterface
    {
        return match (true) {
            str_contains($item->name, self::AGED_BRIE_ITEM) => new AgedBrieItem(),
            str_contains($item->name, self::BACKSTAGE_ITEM) => new BackstagePassItem(),
            str_contains($item->name, self::SULFURAS_ITEM) => new SulfurasItem(),
            str_contains($item->name, self::CONJURED_ITEM) => new ConjuredItem(),
            default => new NormalItem()
        };
    }
}
